### PR TITLE
[Admin][Address] Log order address changes

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | |
| License         | MIT |

history-button
history

With proposed architecture, it should be easy to implement logging functionalities for other entities with their custom, own table 🐃

### CODE REVIEW

File: src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/AddressLogEntry.orm.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        <id name="id" column="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+
+        <field name="action" type="string" />
+        <field name="loggedAt" type="datetime" />
+        <field name="objectId" length="64" nullable="true" />
+        <field name="objectClass" type="string" />
+        <field name="version" type="integer" />
+        <field name="data" type="array" />
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/AddressLogEntry.orm.xml]: Can we use json here?

Comment (Zales0123) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/AddressLogEntry.orm.xml]: Probably 👍

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml [hunk 2]
```diff
@@ -102,3 +108,10 @@ twig:
     exception_controller: "FOS\\RestBundle\\Controller\\ExceptionController::showAction"
     globals:
         sylius: "@sylius.context.shopper"
+
+sylius_resource:
+    resources:
+        sylius.address_log_entry:
+            classes:
+                model: Sylius\Component\Addressing\Model\AddressLogEntry
+                repository: Sylius\Bundle\ResourceBundle\Doctrine\ORM\ResourceLogEntryRepository
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml]: Shouldn't this be specified in `AddressingBundle`?

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/order.yml [hunk 3]
```diff
@@ -21,6 +21,15 @@ sylius_admin_order_show:
...
             permission: true
 
+sylius_admin_order_history:
+    path: /orders/{id}/history
+    defaults:
+        _controller: sylius.controller.order:showAction
+        _sylius:
+            section: admin
+            template: SyliusAdminBundle:Order:history.html.twig
+            permission: true
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/order.yml]: Do we need this?

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/partial/address.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+sylius_admin_address_log_entry_index:
+    path: /log-entry/{objectId}
+    methods: [GET]
+    defaults:
+        _controller: sylius.controller.address_log_entry:indexAction
+        _sylius:
+            template: "@SyliusAdmin/Order/History/addressIndex.html.twig"
+            grid: sylius_admin_address_log_entry
+            section: admin
+            permission: true
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/partial/address.yml]: Do we need this?

File: src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/AddressLogEntry.orm.xml (hunk 1 above)
Comment (Zales0123) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/AddressLogEntry.orm.xml]: Ok, apparently we can't, it causes billions of errors :/

File: src/Sylius/Behat/Page/Admin/Order/HistoryPage.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public function getRouteName()
+    {
+        return 'sylius_admin_order_history';
+    }
+
+    /**
+     * @return int
+     */
+    public function countChanges()
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Order/HistoryPage.php]: `countShippingAddressChanges()`?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/logData.html.twig [hunk 6]
```diff
@@ -0,0 +1 @@
+{% for key, field in data if field is not null %}{{ key }}: {{ field }} {% endfor %}
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/logData.html.twig]: Can we render it as a table?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_actions.html.twig [hunk 7]
```diff
@@ -1,7 +1,8 @@
 {% import '@SyliusUi/Macro/buttons.html.twig' as buttons %}
 
 {% if sm_can(order, 'cancel', 'sylius_order') %}
-    <form action="{{ path('sylius_admin_order_cancel', {'id': order.id} ) }}" method="post">
+    <a href="{{ path('sylius_admin_order_history', {'id': order.id}) }}" class="ui violet labeled icon button" style="float: right;"><i class="history icon"></i>{{ 'sylius.ui.order_history'|trans }}</a>
+    <form action="{{ path('sylius_admin_order_cancel', {'id': order.id} ) }}" method="post" style="float: right;">
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_actions.html.twig]: `{'id': order.id} )` -> `{'id': order.id})`

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/history.html.twig [hunk 8]
```diff
@@ -0,0 +1,29 @@
+{% extends 'SyliusAdminBundle::layout.html.twig' %}
+
+{% import '@SyliusUi/Macro/labels.html.twig' as label %}
+
+{% block title %}{{ parent() }} {{ 'sylius.ui.order'|trans ~' #'~ order.number }} - {{ 'sylius.ui.order_history'|trans }}{% endblock %}
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/history.html.twig]: `{{ 'sylius.ui.order'|trans ~' #'~ order.number }}` -> `{{ 'sylius.ui.order'|trans }} #{{ order.number }}` WDYT?

File: src/Sylius/Component/Grid/Provider/ArrayGridProvider.php [hunk 9]
```diff
@@ -48,7 +48,7 @@ public function get($code)
             throw new UndefinedGridException($code);
         }
 
-        return $this->grids[$code];
+        return clone $this->grids[$code];
```
Comment (pamil) [inline src/Sylius/Component/Grid/Provider/ArrayGridProvider.php]: Adding a comment why might be useful.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/logData.html.twig (hunk 6 above)
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/logData.html.twig]: We could, however it's only a cell in a table which does not contains so much data. I just don't like how table looks, when only one cell is massive and others are have one-line text 😕  🐃

Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/logData.html.twig]: Hmm, so let's just put a new line between each key, so:

```
firstName: Krzysztof lastName: Krawczyk
```

becomes

```
firstName: Krzysztof
lastName: Krawczyk
```

File: src/Sylius/Behat/Page/Admin/Order/HistoryPage.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+namespace Sylius\Behat\Page\Admin\Order;
+
+use Sylius\Behat\Page\SymfonyPage;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+final class HistoryPage extends SymfonyPage implements HistoryPageInterface
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Admin/Order/HistoryPage.php]: Pages should not be final

File: src/Sylius/Bundle/AddressingBundle/Resources/config/app/config.yml [hunk 11]
```diff
@@ -7,3 +7,10 @@ jms_serializer:
             sylius-addressing:
                 namespace_prefix: "Sylius\\Component\\Addressing"
                 path: "@SyliusAddressingBundle/Resources/config/serializer"
+
+sylius_resource:
+    resources:
+        sylius.address_log_entry:
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AddressingBundle/Resources/config/app/config.yml]: Won't it be better to have it as a part of address configuration tree?

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 12]
```diff
@@ -210,6 +211,21 @@ public function theCustomerAddressedItTo(AddressInterface $address)
     }
 
     /**
+     * @When the customer changed shipping address' street to :street
```
Comment (Arminek) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: `@Given`.

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 13]
```diff
@@ -210,6 +211,21 @@ public function theCustomerAddressedItTo(AddressInterface $address)
...
+     */
+    public function theCustomerChangedShippingAddressStreetTo($street)
+    {
+        $order = $this->sharedStorage->get('order');
+        $this->applyTransitionOnOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_ADDRESS);
+
+        /** @var Order $order */
```
Comment (Arminek) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: `OrderInterface`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php [hunk 14]
```diff
@@ -52,6 +53,11 @@
     private $updatePage;
 
     /**
+     * @var HistoryPageInterface
+     */
+    private $historyPage;
```
Comment (Arminek) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php]: It is just a thought but maybe it would be a cool idea to create `ManagingOrdersHistoryContext` WDYT?

File: src/Sylius/Bundle/ResourceBundle/Model/ResourceLogEntry.php [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Sylius\Bundle\ResourceBundle\Model;
+
+use Gedmo\Loggable\Entity\MappedSuperclass\AbstractLogEntry;
+use Sylius\Component\Resource\Model\ResourceInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+abstract class ResourceLogEntry extends AbstractLogEntry implements ResourceInterface
```
Comment (Arminek) [inline src/Sylius/Bundle/ResourceBundle/Model/ResourceLogEntry.php]: Do we really need this abstract class? Probably yes 😢 🐃 . Besides it is strictly coupled with `orm`.

File: src/Sylius/Component/Addressing/Model/AddressLogEntry.php [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Sylius\Component\Addressing\Model;
+
+use Sylius\Bundle\ResourceBundle\Model\ResourceLogEntry;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+final class AddressLogEntry extends ResourceLogEntry
```
Comment (Arminek) [inline src/Sylius/Component/Addressing/Model/AddressLogEntry.php]: 🐃  Maybe we should point `ResourceLogEntry` as an address log entry in configuration, cuz this class has no sense to me it is empty and final 🌵.

Comment (Zales0123) [inline src/Sylius/Component/Addressing/Model/AddressLogEntry.php]: Unfortunately we must have this class, to have custom table for each loggable entity (we don't want to store, for example, product changes in the same table as address changes).

File: src/Sylius/Bundle/ResourceBundle/Model/ResourceLogEntry.php (hunk 15 above)
Comment (Zales0123) [inline src/Sylius/Bundle/ResourceBundle/Model/ResourceLogEntry.php]: There is different gedmo mapping for *orm* and *odm* so I don't think it's a big problem, especially that we don't really support *odm* for now very well 😄

File: src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php (hunk 14 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php]: Would be a good idea, but I think it can be implemented when we will have more things to show in history (for now there is only one page with one method).

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 17]
```diff
@@ -210,6 +210,21 @@ public function theCustomerAddressedItTo(AddressInterface $address)
     }
 
     /**
+     * @Given the customer changed shipping address' street to :street
+     */
+    public function theCustomerChangedShippingAddressStreetTo($street)
+    {
+        $order = $this->sharedStorage->get('order');
+        $this->applyTransitionOnOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_ADDRESS);
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: We should first modify the order, then apply the transition.

File: features/order/managing_orders/modifying_billing_address/modifying_shipping_address_on_order_with_taxes.feature [hunk 18]
```diff
@@ -25,6 +25,6 @@ Feature: Modifying a customer's billing address on an order with taxes
         And I specify their new billing address as "Los Angeles", "Seaside Fwy", "90802", "United States" for "Lucifer Morningstar"
         And I save my changes
         Then I should be notified that it has been successfully edited
-        And this order should be shipped to "Lucifer Morningstar", "Seaside Fwy", "90802", "Los Angeles", "United States"
+        And the order should be billed to "Lucifer Morningstar", "Seaside Fwy", "90802", "Los Angeles", "United States"
```
Comment (pjedrzejewski) [inline features/order/managing_orders/modifying_billing_address/modifying_shipping_address_on_order_with_taxes.feature]: I think naming of this features is wrong, cause we have both billing and shipping here. Can you fix it, please? Otherwise these changes are quite confusing.

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/order.yml [hunk 19]
```diff
@@ -19,7 +19,14 @@ sylius_admin_order_show:
         _sylius:
             section: admin
             template: SyliusAdminBundle:Order:show.html.twig
-            permission: true
+
+sylius_admin_order_history:
+    path: /orders/{id}/history
+    defaults:
+        _controller: sylius.controller.order:showAction
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/order.yml]: Good solution!

File: src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/action.html.twig [hunk 20]
```diff
@@ -0,0 +1,5 @@
+{% if data == 'create' %}
+    <span class="ui green label">{{ data }}</span>
+{% elseif data == 'update' %}
+    <span class="ui yellow label">{{ data }}</span>
+{% endif %}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Address/Grid/Field/action.html.twig]: This and template below could go to UiBundle, like other generic fields.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/History/addressIndex.html.twig [hunk 21]
```diff
@@ -0,0 +1,24 @@
+{% import 'SyliusUiBundle:Macro:table.html.twig' as table %}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/History/addressIndex.html.twig]: This template is a partial, so should be ``@SyliusAdmin/Order/History/_address.html.twig``.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_actions.html.twig [hunk 22]
```diff
@@ -1,7 +1,8 @@
 {% import '@SyliusUi/Macro/buttons.html.twig' as buttons %}
 
 {% if sm_can(order, 'cancel', 'sylius_order') %}
-    <form action="{{ path('sylius_admin_order_cancel', {'id': order.id} ) }}" method="post">
+    <a href="{{ path('sylius_admin_order_history', {'id': order.id}) }}" class="ui violet labeled icon button" style="float: right;"><i class="history icon"></i>{{ 'sylius.ui.order_history'|trans }}</a>
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Order/Show/_actions.html.twig]: This link should be outside of the ``if`` otherwise it will be hidden for cancelled orders. ;)

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml [hunk 23]
```diff
@@ -29,6 +29,12 @@ doctrine:
         entity_managers:
             default:
                 auto_mapping: true
+                mappings:
+                    gedmo_loggable:
+                        type: annotation
+                        prefix: Sylius\Bundle\ResourceBundle\Model
+                        dir: "%kernel.root_dir%/../src/Sylius/Bundle/ResourceBundle/Model"
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml]: This won't work in Sylius-Standard.

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml]: Yes :/ Do you think we should provide this path as parameters? Or override this configuration in Sylius-Standard `app/config.yml`?

File: src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php (hunk 14 above)
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php]: Can be a case for separate PR, but I would say to do it right now. I am afraid that we will forget about it and end up with too big class

File: src/Sylius/Bundle/AddressingBundle/DependencyInjection/Configuration.php [hunk 24]
```diff
@@ -90,6 +92,21 @@ private function addResourcesSection(ArrayNodeDefinition $node)
...
+                            ->addDefaultsIfNotSet()
+                            ->children()
+                                ->variableNode('options')->end()
+                                ->arrayNode('classes')
+                                    ->addDefaultsIfNotSet()
+                                    ->children()
+                                        ->scalarNode('model')->defaultValue(AddressLogEntry::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('controller')->defaultValue(ResourceController::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('repository')->defaultValue(ResourceLogEntryRepository::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('factory')->defaultValue(Factory::class)->end()
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AddressingBundle/DependencyInjection/Configuration.php]: I'm wondering why don't we have `->cannotBeEmpty()` on a factory node also. Any thoughts?

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! 👍

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml (hunk 23 above)
Comment (tchapi) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml]: Indeed, it breaks every project using Sylius as a dependency, which is a problem (https://github.com/Sylius/Sylius/issues/7156). Yet it was merged ?
